### PR TITLE
Fix unittest to work with cherrypy 3.2.2

### PR DESCRIPTION
**Pull request by vkuznet**

### CODE REVIEW

File: test/python/WMCore_t/REST_t/Api_t.py [hunk 1]
```diff
@@ -225,9 +225,23 @@ def test_multi_throw10(self):
...
 def setup_server():
     srcfile = __file__.split("/")[-1].split(".py")[0]
-    setup_test_server(srcfile, "Root")
-    #cpconfig.update({"log.screen": True})
+    server, T.test_authz_key = setup_test_server(srcfile, "Root")
+    print("SERVER", server.config.pythonise_())
+    print("AUTHZ", T.test_authz_key, T.test_authz_key.data)
+
+import cherrypy
+from cherrypy.test import webtest
```
Comment (ericvaandering) [inline test/python/WMCore_t/REST_t/Api_t.py]: Let's keep imports at the top unless this is just a test Valentin is trying to run (not really asking to merge)

Comment (vkuznet) [inline test/python/WMCore_t/REST_t/Api_t.py]: Yes, I know I'll make everything clean once I'll see results. I also
plant to re-factor a little bit and move this section into module which will be
used by other tests. That's why I kept them together (the imports and
underneath code which will be moved).

Comment (vkuznet) [inline test/python/WMCore_t/REST_t/Api_t.py]: Eric,
I think something really screwed on bot environment. From this test:
https://cmssdt.cern.ch/jenkins/job/DMWM-PR-test/442/testReport/junit/(root)/Api_t/testmain/

I see the following:

/build/cmsbuild/jenkins-workarea/workspace/DMWM-PR-unittests/SLICE/1/label/slc6_amd64-cloud/deploy/1.0.9.pre9/sw.pre/slc6_amd64_gcc481/external/cherrypy/3.2.2-comp10/lib/python2.6/site-packages/cherrypy/process/servers.py",
line 187, in _start_http_thread     self.httpserver.start() AttributeError:
'str' object has no attribute 'start'

it means that someone overwrite self.httpserver which now is a string rather
then cherrypy object.

I built cherrypy in my VM, so I would assume I have similar environment and I
can run unit tests just fine. But on a bot it is different.

So, I want to see bot installation and environment. Do we have access to the
node that we can manually run the unittest from there within bot environment?

**Comment by ericvaandering** [comment]
Hi @vkuznet , I don't know about logging into these nodes, you can ask Shahzad, but I doubt that gets you anything since we install our software stack from scratch to run the tests. So nothing on the node itself should be causing an issue like this.

What might help you is this: 
```bash
set -x

echo "Job name is"
echo $JOB_NAME

#git config --global http.postBuffer 524288000

rm -rf cms-bot || true
if [ ! -d cms-bot ]; then
  git clone https://github.com/$CMS_BOT_REPO/cms-bot
fi

pushd $WORKSPACE/cms-bot/

wget https://pypi.python.org/packages/source/r/requests/requests-2.3.0.tar.gz#md5=7449ffdc8ec9ac37bbcd286003c80f00
tar -xvf requests-2.3.0.tar.gz
rm -rf requests || true
mv requests-2.3.0/requests/ requests

git checkout $CMS_BOT_BRANCH
git pull --rebase origin $CMS_BOT_BRANCH

popd

. ./cms-bot/DMWM/kill-databases.sh

. ./cms-bot/DMWM/setup-secrets.sh
. ./cms-bot/DMWM/update-deployment.sh
. ./cms-bot/DMWM/update-wmcore.sh

pushd code
git fetch --tags  https://github.com/dmwm/WMCore.git "+refs/heads/*:refs/remotes/origin/*"
git config remote.origin.url https://github.com/dmwm/WMCore.git
git config --add remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"
git fetch --tags  https://github.com/dmwm/WMCore.git "+refs/pull/*:refs/remotes/origin/pr/*"
export COMMIT=`git rev-parse "origin/pr/$ghprbPullId/merge^{commit}"`
(git checkout master && git merge $COMMIT) || git checkout -f $COMMIT
popd 


#. ./cms-bot/DMWM/temporary-patches.sh

. ./cms-bot/DMWM/latest-dmwm-versions.sh

if [ -z "$WMAGENT_VERSION" ]; then
  export WMAGENT_VERSION=$WMAGENT_LATEST
fi

. ./cms-bot/DMWM/deploy-wmagent.sh
. ./cms-bot/DMWM/activate-start-agent.sh
. ./cms-bot/DMWM/install-wmcore.sh
export PYTHONPATH=`pwd`/code/test/python:`pwd`/code/src/python:$PYTHONPATH

. ./cms-bot/DMWM/test-wmcore.sh

echo "Stopping WMAgent and services"
$PWD/deploy/current/config/wmagent/manage stop-agent || true
$PWD/deploy/current/config/wmagent/manage stop-services || true
    
. ./cms-bot/DMWM/kill-databases.sh
``` 
That's the script that runs all the unit tests. You'll find various environment variables scattered around and the scripts that do the real work are in https://github.com/cms-sw/cms-bot. The thing you really want to concentrate on, I think, is https://github.com/cms-sw/cms-bot/blob/master/DMWM/test-wmcore.sh#L34 where the tests are actually run. $SLICES is 10. Let me try to get back to you with the actual $SLICE setting to get this particular test.

**Comment by ericvaandering** [comment]
Actually you can see SLICE=1 from what you posted in the code comments. And just about any SLICE will have some Api_t test in it.

**Comment by vkuznet** [comment]
Eric, Seangchan,
I was able to reproduce unit test failures using bot script provided by Eric. Basically I only needed one command:
```
python setup.py test --buildBotMode=true --reallyDeleteMyDatabaseAfterEveryTest=true --testCertainPath=test/python --testTotalSlices=10 --testCurrentSlice=1
```

Using this command I got identical failures and I spent couple of days to figure out what's wrong. Turns out WMCore_t tests have multiple issues, e.g. global variables, cherrypy bus redefinition, obsolete code and possible state redefinition of cherrypy server. It is total mess and require serious attention. I want to commit another set of changes which fixed issues with unit tests in my local setup. Please don't merge them yet since I want to see results from the build. The code I'm about to submit did the following: remove global authz key file, define port for cherrypy, define local server within separate process and switch to webtest based class. I'll push my changes to fix_api_t2 branch.

**Comment by ericvaandering** [comment]
I'm sure. Are the problems with setup.py or with the test/ dir? I'll welcome any fixes but would be good to understand if those things were done strangely.

Sent from a mobile device. Please excuse my brevity or transcription errors.

On Sep 1, 2015, at 18:15, Valentin Kuznetsov <notifications@github.com<mailto:notifications@github.com>> wrote:


Eric, Seangchan,
I was able to reproduce unit test failures using bot script provided by Eric. Basically I only needed one command:

python setup.py test --buildBotMode=true --reallyDeleteMyDatabaseAfterEveryTest=true --testCertainPath=test/python --testTotalSlices=10 --testCurrentSlice=1


Using this command I got identical failures and I spent couple of days to figure out what's wrong. Turns out WMCore_t tests have multiple issues, e.g. global variables, cherrypy bus redefinition, obsolete code and possible state redefinition of cherrypy server. It is total mess and require serious attention. I want to commit another set of changes which fixed issues with unit tests in my local setup. Please don't merge them yet since I want to see results from the build. The code I'm about to submit did the following: remove global authz key file, define port for cherrypy, define local server within separate process and switch to webtest based class. I'll push my changes to fix_api_t2 branch.

**Comment by vkuznet** [comment]
Eric,
I'm not sure how to read this, since from the web page I see that test
https://cmssdt.cern.ch/jenkins/job/DMWM-PR-test/450/testReport/Api_t/Tester/test_accept_good_json/
has Failed. While the Autobot said
Api_t.Tester:test_simple_json changed from error to success
Am I missing something?
Could you please clarify which message we should trust.

**Comment by ericvaandering** [comment]
The reason for that is that test was already failing in the last test of the baseline: https://cmssdt.cern.ch/jenkins/view/DMWM/job/DMWM-WMCore-UnitTests/lastCompletedBuild/testReport/Api_t/Tester/test_accept_good_json/ The failures do look different and I don't try to detect that. I also don't try to detect which unit tests you might have tried to fix.

In the past we had a static list of unit tests that were allowed to fail, but now we have so many with no "good" reason, that I decided to do something different and only record changes.

But, it does look like I will owe you some vodka. :-) October at CERN?

**Comment by ericvaandering** [comment]
Ah, different question that you asked. The test you quote is different than the one that changed to success. Even easier answer. :-)

**Comment by vkuznet** [comment]
I was exhausted by failures and didn't catch that indeed it was a different
test.

October drink(s) is scheduled in my calendar.

**Comment by ticoann** [comment]
Thanks Valentin, well deserved for getting wasted in October. :smile:

**Comment by vkuznet** [comment]
Eric,
how I can view full log from BuildBot? So far I can't see my logger messages
from cherrypy which go to stdout. I bet there is a log with all stdout/stderr
messages but web interface does not show it.
V.

**Comment by ericvaandering** [comment]
Was this supposed to be merged?

**Comment by vkuznet** [comment]
Seangchan,
I asked to post-pone the merge, since I still play around with this PR.
Oh well, it's too late, I need to create another PR.

**Comment by ticoann** [comment]
Was not? Sorry, I was misunderstood I thought this was done since some unittest are fixed. Sorry about that. I merged since it seems it wouldn't affect any other part than test.

**Comment by ticoann** [comment]
I can revert if you prefer.

**Comment by ericvaandering** [comment]
I think so. Valentin can then push new commits onto this.

**Comment by ticoann** [comment]
OK. Valentin, is it OK if I revert the merge. You would have to create the new pull request.

**Comment by hufnagel** [comment]
Why ? If you revert the merge doesn't the old pull request stay active ? (never done it)

**Comment by ticoann** [comment]
No, It seems github 'revert' makes another commit on top of the merged code automatically. (which make sense otherwise it will mess up commit history)

**Comment by vkuznet** [comment]
Ok, we can discard this PR. I created a new one
https://github.com/dmwm/WMCore/pull/6191

I explicitly ask do not merge it until we fully understand the problem.

**Pull request merged**